lib/inc/drogon/utils/HttpClientPool.h

```diff
@@ -0,0 +1,321 @@
+        std::shared_ptr<trantor::EventLoopThreadPool> dispatchPool = nullptr)
+        : cfg_(std::move(cfg)),
+          loopPool_(std::move(loopPool)),
+          dispatchPool_(std::move(dispatchPool))
+    {
+        if (loopPool_ == nullptr)
+        {
+            loopPool_ = std::make_shared<trantor::EventLoopThreadPool>(
+                cfg_.numOfThreads);
+            loopPool_->start();
+        }
+        if (dispatchPool_ == nullptr)
+        {
+            dispatchPool_ = std::make_shared<trantor::EventLoopThreadPool>(1);
+            dispatchPool_->start();
+        }
+        loopPtr_ = dispatchPool_->getNextLoop();
+
+        for (std::size_t i = 0; i < cfg_.size; i++)
+        {
+            auto loopPtr = loopPool_->getNextLoop();
+            auto func = [this, loopPtr]() mutable {
+                auto client = HttpClient::newHttpClient(cfg_.hostString,
+                                                        loopPtr,
+                                                        cfg_.useOldTLS,
+                                                        cfg_.validateCert);
+                if (cfg_.setCallback)
+                    cfg_.setCallback(client);
+                return client;
+            };
+            httpClients_.emplace(std::make_shared<Connection>(func, cfg_));
+        }
+        LOG_DEBUG << "httpClients_ size:" << httpClients_.size();
+
+        if (cfg_.idleTimeout.has_value() && cfg_.checkInterval.has_value())
+        {
+            timerId_ = loopPtr_->runEvery(cfg_.checkInterval.value(), [this] {
+                std::unique_lock<std::mutex> lock(mutex_);
+                if (httpClients_.empty())
+                    return;
+                std::queue<std::shared_ptr<Connection>> clients;
+                while (!httpClients_.empty())
+                {
+                    auto connPtr = std::move(httpClients_.front());
+                    httpClients_.pop();
+                    if (connPtr->reachIdleTimeout())
+                    {
+                        // close tcp connection
+                        connPtr->resetClientPtr();
+                    }
+                    clients.emplace(std::move(connPtr));
+                }
+                httpClients_ = std::move(clients);
+            });
+        }
+    }
+
+    ~HttpClientPool()
+    {
+        if (timerId_)
+        {
+            std::promise<void> done;
+            loopPtr_->runInLoop([&] {
+                loopPtr_->invalidateTimer(timerId_.value());
+                done.set_value();
+            });
+            done.get_future().wait();
+        }
+        [](auto &&...args) {
+            auto func = [](auto &poolPtr) {
+                if (poolPtr == nullptr)
+                    return;
+                for (auto &ptr : poolPtr->getLoops())
+                    ptr->runInLoop([=] { ptr->quit(); });
+                poolPtr->wait();
+                poolPtr.reset();
+            };
+            (func(args), ...);
```

Comment:
The thread pool may not created by HttpClientPool,  we should not quit thread pool unconditionally.

Reply:
update